providers/common/ciphers/gcm.c

```diff
@@ -0,0 +1,586 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/evp.h>
+#include <openssl/params.h>
+#include <openssl/core_numbers.h>
+#include <openssl/core_names.h>
+#include "internal/rand_int.h"
+#include "internal/provider_algs.h"
+#include "internal/provider_ctx.h"
+#include "internal/providercommonerr.h"
+#include "ciphers_locl.h"
+
+/* TODO(3.0) Figure out what flags are really needed */
+#define AEAD_GCM_FLAGS (EVP_CIPH_FLAG_AEAD_CIPHER | EVP_CIPH_FLAG_DEFAULT_ASN1 \
+                       | EVP_CIPH_CUSTOM_IV | EVP_CIPH_FLAG_CUSTOM_CIPHER      \
+                       | EVP_CIPH_ALWAYS_CALL_INIT | EVP_CIPH_CTRL_INIT        \
+                       | EVP_CIPH_CUSTOM_COPY)
+
+static OSSL_OP_cipher_encrypt_init_fn gcm_einit;
+static OSSL_OP_cipher_decrypt_init_fn gcm_dinit;
+static OSSL_OP_cipher_ctx_get_params_fn gcm_ctx_get_params;
+static OSSL_OP_cipher_ctx_set_params_fn gcm_ctx_set_params;
+static OSSL_OP_cipher_cipher_fn gcm_cipher;
+static OSSL_OP_cipher_update_fn gcm_stream_update;
+static OSSL_OP_cipher_final_fn gcm_stream_final;
+
+static int gcm_tls_init(PROV_GCM_CTX *dat, unsigned char *aad, size_t aad_len);
+static int gcm_tls_iv_set_fixed(PROV_GCM_CTX *ctx, unsigned char *iv,
+                                size_t len);
+static int gcm_tls_cipher(PROV_GCM_CTX *ctx, unsigned char *out, size_t *padlen,
+                          const unsigned char *in, size_t len);
+static int gcm_cipher_internal(PROV_GCM_CTX *ctx, unsigned char *out,
+                               size_t *padlen, const unsigned char *in,
+                               size_t len);
+
+static void gcm_initctx(void *provctx, PROV_GCM_CTX *ctx, size_t keybits,
+                        const PROV_GCM_HW *hw, size_t ivlen_min)
+{
+    ctx->pad = 1;
+    ctx->mode = EVP_CIPH_GCM_MODE;
+    ctx->taglen = -1;
+    ctx->tls_aad_len = -1;
+    ctx->ivlen_min = ivlen_min;
+    ctx->ivlen = (EVP_GCM_TLS_FIXED_IV_LEN + EVP_GCM_TLS_EXPLICIT_IV_LEN);
+    ctx->keylen = keybits / 8;
+    ctx->hw = hw;
+    ctx->libctx = PROV_LIBRARY_CONTEXT_OF(provctx);
+}
+
+static void gcm_deinitctx(PROV_GCM_CTX *ctx)
+{
+    OPENSSL_cleanse(ctx->iv, sizeof(ctx->iv));
+}
+
+/*
+ * Note: The passed in ivlen is ignored. since the default is already set OR
+ * the user has set the required ivlen.
+ */
+
+/* TODO(3.0) Figure out why we pass in default values here that are not used */
```

Comment:
I think this really needs resolving before we can merge this.

The code in `EVP_CipherInit_ex` that passes the key length/iv length is new with providers - and this is the first time that we have moved a cipher into the providers with a variable iv length.

`EVP_CipherInit_ex` is using `EVP_CIPHER_CTX_iv_length()` - which sounds like it should return the *actual* iv length that has been configured. It seems like a bug to me that it doesn't do this for GCM (and if so we should backport the fix to 1.1.1). Alternatively we're going to need a new function that does get the actual IV length.

Reply:
This looks like a misunderstanding.  The passed key length and iv length aren't meant to be some kind of default, but the actual length of the passed `key` and `iv`.  They should be verified against the key and iv lengths that the implementation expects.

Unfortunately, the EVP API doesn't allow callers to say what size the diverse data they pass in have, so the EVP API must make assumptions, and the best assumption it knows for the IV is the passed data has the correct size.  This *is* dicy, but I've insisted in the libcrypto\<-\>provider interface to specify buffer sizes more or less everywhere, to allow for a future API with better size control.

Reply:
Ok so maybe it should pass -1 in instead? Even if I add another function to get the iv it is still pointless. Since it will retrieve its own internal value and then set the iven to this value.. which seems rather pointless.

Reply:
Ok I have updated the code (no documentation yet). I dont know what to call the iv function.

Reply:
No, I'm sorry, that's a bad idea.  Basically, I would say that for algorithms with variable IV size, the current `EVP_CIPHER_CTX_iv_length()` code is simply *wrong*, i.e. it's ill fitted for that sort of thing.  It's only because we have somehow been able to ignore that function when passing an IV to aria (and the application was otherwise *suppose to know* what IV size it has requested) that we haven't noticed.

So basically, `EVP_CIPHER_CTX_iv_length() should do exactly what you implemented as `EVP_CIPHER_CTX_cur_iv_length()`

Reply:
I'm sorry, I worded this wrong.  Your analysis and solution is actually correct, but should come as a correction of `EVP_CIPHER_CTX_iv_length()` rather than a new function.

Reply:
EVP_CIPHER_CTX_iv_length() has been changed to return the ctx->ivlen.
EVP_CIPHER_CTX_cur_iv_length(0 has been removed..